MATLAB/update_FC_state.m

```diff
@@ -12,91 +12,149 @@
-        else
-            state.tumbling_trigger_count = 0;
-        end
-        if (state.tumbling_trigger_count>=2)
-            %tumbling triggered twice in a row
-            state.adcs_state='tumbling';
-            state.detumbled_trigger_count = 0;
-            state.detumbler_state=0;
-            state.magrod_moment_cmd= [0;0;0;];
-        end
-    otherwise
-        actuators.magrod_moment= [0;0;0;];
-        actuators.wheel_torque = [100;100;100;];
-        actuators.wheel_enable=[0;0;0;];
+previous_main_state=state.main_state;
+if strcmp(state.main_state,'off')
+    state.main_state='hardware setup';
+end
+if strcmp(state.main_state,'hardware setup')
+    %move to next state if all sub systems are setup
+    if true
+        state.on_time=uint64(0);
+        state.main_state='initialization hold';
+    end 
+end
+if strcmp(state.main_state,'initialization hold')
+    %move to next state if initialization time has passed
+    if state.initialization_hold_done || state.on_time>(30*60*1E9)
+        state.initialization_hold_done=true;%save this in eprom
+        state.main_state='detumble';
+    end 
+end
+if strcmp(state.main_state,'detumble')
+    %detumble
+    if norm(estimator.angular_momentum_body)<const.MAXWHEELRATE*const.JWHEEL*0.2
+        state.main_state='get gps';
+    end 
+end
+if strcmp(state.main_state,'get gps')
+    %check if gps is valid
+    if estimator.tumbling==1 || estimator.low_power==1
+        state.main_state='detumble';
+    elseif estimator.valid_gps
+        state.main_state='get target orbit';
+    end
+    %calculate pointing strategy
+    state.primary_current_direction_body= estimator.sat2sun_body/norm(estimator.sat2sun_body);
+    state.primary_desired_direction_body= [1;0;0];
+    state.secondary_current_direction_body=NaN(3,1);
+    state.secondary_desired_direction_body=NaN(3,1);
+end
+if strcmp(state.main_state,'get target orbit')
+    if estimator.tumbling || estimator.low_power || estimator.eclipse || ~estimator.valid_gps
+        state.main_state='detumble';
+    elseif estimator.valid_target_gps
+        state.main_state='rendezvous';
+    end 
+    %calculate pointing strategy
+    r_hat_eci= estimator.position_eci/norm(estimator.position_eci);
+    v_hat_eci= estimator.velocity_eci/norm(estimator.velocity_eci);
+    r_hat_body= utl_rotateframe(estimator.quat_body_eci,r_hat_eci);
+    v_hat_body= utl_rotateframe(estimator.quat_body_eci,v_hat_eci);
+    r_hat_cross_sat2sun_body=cross(estimator.sat2sun_body,r_hat_eci);
+    state.primary_current_direction_body= r_hat_body;
+    state.primary_desired_direction_body= [1;0;0];
+    state.secondary_current_direction_body= r_hat_cross_sat2sun_body/norm(r_hat_cross_sat2sun_body);
+    state.secondary_desired_direction_body= [0;0;1];
+end
+if strcmp(state.main_state,'rendezvous')
+    if estimator.tumbling || estimator.low_power || estimator.eclipse || ~estimator.valid_gps || ~estimator.valid_target_gps
```

Comment:
This may be temporary, but if everything is behaving nominally except for having target position, velocity would we want to go back to detumble or hold attitude with the full ADCS system? The latter would probably give us a better shot at getting comms, gps, etc

Reply:
Hopefully if cd gps lock is lost, the orbit estimator can just propagate the target orbit, that way it won't have to go into detumble.